### ISSUE TITLE
Uncaught TypeError: Cannot read property 'length' of undefined

### ISSUE THREAD

**Issue by onlylin** [atom/sort-lines#39]
[Enter steps to reproduce below:]

1. ...
2. ...

**Atom Version**: 1.0.19
**System**: Mac OS X 10.10.5
**Thrown From**: [sort-lines](https://github.com/atom/sort-lines) package, v0.11.0

### Stack Trace

Uncaught TypeError: Cannot read property 'length' of undefined

```
At /Users/Lin/.atom/packages/sort-lines/lib/range-finder.coffee:38

TypeError: Cannot read property 'length' of undefined
  at RangeFinder.module.exports.RangeFinder.sortableRangeFrom (/Users/Lin/.atom/packages/sort-lines/lib/range-finder.coffee:38:50)
  at RangeFinder.module.exports.RangeFinder.ranges (/Users/Lin/.atom/packages/sort-lines/lib/range-finder.coffee:16:9)
  at Function.module.exports.RangeFinder.rangesFor (/Users/Lin/.atom/packages/sort-lines/lib/range-finder.coffee:7:29)
  at sortTextLines (/Users/Lin/.atom/packages/sort-lines/lib/sort-lines.coffee:20:32)
  at sortLines (/Users/Lin/.atom/packages/sort-lines/lib/sort-lines.coffee:27:4)
  at atom-text-editor.atom.commands.add.sort-lines:sort (/Users/Lin/.atom/packages/sort-lines/lib/sort-lines.coffee:8:9)
  at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/Applications/Atom.app/Contents/Resources/app.asar/src/command-registry.js:245:29)
  at /Applications/Atom.app/Contents/Resources/app.asar/src/command-registry.js:3:61
  at CommandPaletteView.module.exports.CommandPaletteView.confirmed (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/command-palette/lib/command-palette-view.js:159:32)
  at CommandPaletteView.module.exports.SelectListView.confirmSelection (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-space-pen-views/lib/select-list-view.js:338:21)
  at space-pen-div.atom.commands.add.core:confirm (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-space-pen-views/lib/select-list-view.js:109:19)
  at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/Applications/Atom.app/Contents/Resources/app.asar/src/command-registry.js:245:29)
  at /Applications/Atom.app/Contents/Resources/app.asar/src/command-registry.js:3:61
  at KeymapManager.module.exports.KeymapManager.dispatchCommandEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:524:16)
  at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:347:22)
  at HTMLDocument.module.exports.WindowEventHandler.onKeydown (/Applications/Atom.app/Contents/Resources/app.asar/src/window-event-handler.js:184:20)

```

### Commands

```
     -1:45.1.0 editor:consolidate-selections (atom-text-editor.editor.mini.is-focused)
     -1:45.1.0 core:cancel (atom-text-editor.editor.mini.is-focused)
     -1:41.4.0 command-palette:toggle (atom-text-editor.editor.is-focused)
  2x -1:21.4.0 core:backspace (atom-text-editor.editor.mini.is-focused)
     -1:17.6.0 core:confirm (atom-text-editor.editor.mini.is-focused)
     -1:17.6.0 styleguide:show (atom-text-editor.editor)
     -0:37.9.0 command-palette:toggle (atom-text-editor.editor.is-focused)
     -0:35.4.0 core:confirm (atom-text-editor.editor.mini.is-focused)
     -0:35.4.0 sort-lines:sort (atom-text-editor.editor)
     -0:22.6.0 editor:consolidate-selections (atom-text-editor.editor.is-focused)
     -0:22.6.0 core:cancel (atom-text-editor.editor.is-focused)
     -0:09.0 command-palette:toggle (atom-text-editor.editor.is-focused)
     -0:06.2.0 core:move-down (atom-text-editor.editor.mini.is-focused)
     -0:03.9.0 core:move-up (atom-text-editor.editor.mini.is-focused)
     -0:00.8.0 core:confirm (atom-text-editor.editor.mini.is-focused)
     -0:00.8.0 sort-lines:sort (atom-text-editor.editor)
```

### Config

```json
{
  "core": {}
}
```

### Installed Packages

```coffee
# User
sort-lines, v0.11.0

# Dev
No dev packages
```

**Comment by izuzak** [comment]
@onlylin Do you remember how you triggered this error? I'm trying to reproduce this, but not succeeding. If you can share some steps to reproduce the problem -- that would be great. Thanks for the report.

**Comment by Tyriar** [comment]
Managed to repro:

1. Open styleguide
2. Select the "Something you typed..." text field
3. sort-lines:sort

You can also trigger it on the command palette:

1. Open styleguide
2. Open the command palette
3. sort-lines:sort

Looks like an undefined check is probably sufficient to fix this.

**Comment by Tyriar** [comment]
Actually what I reproduced is captured in https://github.com/atom/sort-lines/issues/29 which was closed without comment.

**Comment by izuzak** [comment]
Thanks for helping investigate this, @Tyriar :bow: and for opening a PR :sparkles:

**Comment by izuzak** [comment]
Should be fixed by https://github.com/atom/sort-lines/pull/42.

**Comment by Tyriar** [comment]
https://github.com/atom/sort-lines/pull/42 probably won't fix this, an exception will still be thrown if `TextEditor.lineTextForBufferRow` returns `undefined` which happens when `endRow` which is passed in as an arg is invalid. Still not sure how to repro that though.

**Comment by izuzak** [comment]
@Tyriar Hmmm, not sure I understand. I was able to reproduce this issue (same stack trace) before #42 landed (by triggering sort lines in a mini editor), and I'm no longer able to after merging #42.

> an exception will still be thrown if TextEditor.lineTextForBufferRow returns undefined which happens when endRow (passed in as an arg) is invalid. Still not sure how to repro that though.

As far as I can tell by stepping through the code, that case will happen when the sort lines call is made with a mini editor focused. In that case, `endRow` ends up being -1, which causes problems. That again makes me think that you did fix this with #42. I can't think of another situation which might cause this, and if it exists -- I think we can wait for someone to confirm that by reporting an exception when it happens. What do you think?

**Comment by Tyriar** [comment]
Are you sure you weren't reproducing #29 which is a distinct trace? But sounds good, let's keep it closed considering that we both have difficulty reproducing this stack trace after #42 change.

**Comment by izuzak** [comment]
> Are you sure you weren't reproducing #29 which is a distinct trace?

Yeah, I believe it's this one -- here's a screenshot:

![screen shot 2015-10-28 at 18 12 27](https://cloud.githubusercontent.com/assets/38924/10796677/84c44b08-7d9f-11e5-8bf6-c930bd943b4e.png)

I opened a new tab, toggled find and replace, put the cursor into the mini-editor for entering the search query, and finally executed "Sort Lines: Sort" from the Command Palette. If the F5 keybinding is used to trigger the sort command, then the bottom part of the stack trace matches as well.

**Comment by Tyriar** [comment]
Great then, fixed 2 :wink: :fireworks:

**Comment by ilikebits** [comment]
I'm still getting this error. Steps to repro:
1. `atom`
2. `ctrl-n` (or otherwise open new buffer)
3. `f5` (or otherwise invoke `sort-lines:sort`)

I'm running Atom 1.3.0 on Ubuntu 14.04.3, with sort-lines v0.12.0.

### Stack Trace

Uncaught TypeError: Cannot read property 'length' of undefined

```
At /home/leo/.atom/packages/sort-lines/lib/range-finder.coffee:44

TypeError: Cannot read property 'length' of undefined
    at RangeFinder.module.exports.RangeFinder.sortableRangeFrom (/home/leo/.atom/packages/sort-lines/lib/range-finder.coffee:44:56)
    at RangeFinder.module.exports.RangeFinder.ranges (/home/leo/.atom/packages/sort-lines/lib/range-finder.coffee:19:22)
    at Function.module.exports.RangeFinder.rangesFor (/home/leo/.atom/packages/sort-lines/lib/range-finder.coffee:8:38)
    at sortTextLines (/home/leo/.atom/packages/sort-lines/lib/sort-lines.coffee:20:32)
    at sortLines (/home/leo/.atom/packages/sort-lines/lib/sort-lines.coffee:27:3)
    at atom-text-editor.atom.commands.add.sort-lines:sort (/home/leo/.atom/packages/sort-lines/lib/sort-lines.coffee:8:9)
    at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/opt/atom/resources/app.asar/src/command-registry.js:260:29)
    at /opt/atom/resources/app.asar/src/command-registry.js:3:61
    at KeymapManager.module.exports.KeymapManager.dispatchCommandEvent (/opt/atom/resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:536:16)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/opt/atom/resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:359:22)
    at WindowEventHandler.module.exports.WindowEventHandler.handleDocumentKeydown (/opt/atom/resources/app.asar/src/window-event-handler.js:99:36)
    at HTMLDocument.<anonymous> (/opt/atom/resources/app.asar/src/window-event-handler.js:3:61)
```

### Commands

```
     -5:48.8.0 core:confirm (atom-text-editor.editor.mini.is-focused)
     -5:48.8.0 sort-lines:sort (atom-text-editor.editor)
     -5:47.6.0 key-binding-resolver:toggle (atom-text-editor.editor.is-focused)
     -5:46.2.0 command-palette:toggle (atom-text-editor.editor.is-focused)
     -5:44.3.0 core:confirm (atom-text-editor.editor.mini.is-focused)
  2x -5:44.3.0 sort-lines:sort (atom-text-editor.editor)
     -0:58.5.0 editor:consolidate-selections (atom-text-editor.editor.is-focused)
     -0:58.5.0 core:cancel (atom-text-editor.editor.is-focused)
     -0:55.7.0 command-palette:toggle (atom-text-editor.editor.is-focused)
     -0:54.4.0 core:select-all (atom-text-editor.editor.mini.is-focused)
  2x -0:53 core:move-down (atom-text-editor.editor.mini.is-focused)
     -0:52.7.0 core:confirm (atom-text-editor.editor.mini.is-focused)
     -0:52.7.0 settings-view:view-installed-packages (atom-text-editor.editor)
  3x -0:49.6.0 core:backspace (atom-text-editor.editor.mini.is-focused)
     -0:01.6.0 core:close (atom-text-editor.editor.mini.is-focused)
     -0:00.7.0 sort-lines:sort (atom-text-editor.editor.is-focused)
```

### Config

```json
{
  "core": {
    "disabledPackages": [
      "metrics",
      "build"
    ],
    "destroyEmptyPanes": false,
    "autoHideMenuBar": true,
    "themes": [
      "one-dark-ui",
      "monokai"
    ],
    "openEmptyEditorOnStart": false,
    "ignoredNames": [
      ".git",
      ".hg",
      ".svn",
      ".DS_Store",
      "._*",
      "Thumbs.db",
      "*.pyc",
      "__pycache__",
      "*.class"
    ]
  }
}
```

### Installed Packages

```coffee
# User
Stylus, v1.2.1
docblockr, v0.7.3
file-icons, v1.6.13
haskell-ghc-mod, v1.2.0
highlight-line, v0.11.0
highlight-selected, v0.11.1
ide-haskell-cabal, v1.0.0
language-cool, v0.4.1
language-haskell, v1.5.1
language-idris, v0.4.0
language-jade, v0.6.2
language-latex, v0.6.1
language-mips, v0.2.0
language-purescript, v0.6.0
language-scala, v1.1.4
linter, v1.11.3
linter-chktex, v1.2.3
linter-clang, v3.3.0
linter-coffeelint, v1.1.2
linter-javac, v1.3.0
linter-jshint, v1.2.1
linter-pylint, v1.2.0
linter-scalac, v1.1.0
minimap, v4.17.0
minimap-find-and-replace, v4.4.0
minimap-git-diff, v4.1.8
minimap-highlight-selected, v4.3.1
minimap-linter, v1.0.0
monokai, v0.18.0
pretty-json, v0.4.1
project-manager, v2.6.5
sort-lines, v0.12.0
symbols-tree-view, v0.10.1

# Dev
No dev packages
```
